Solution

vladyslav-tmf opened the pull request:
![Screenshot_1](https://github.com/user-attachments/assets/3ed50903-0a25-42db-b26a-4529aef6f96b)

vladyslav-tmf committed:
Add Docker setup and weather API integration

- Introduced a Dockerfile to containerize the Python application.
- Configured a .dockerignore file to exclude unnecessary files from the Docker context.
- Implemented get_weather function in main.py to fetch and display current weather data for Paris using Weather API.
- Added a new dependency into requirements.txt file.
- Added .env.sample file.
- Modified .gitignore.

vladyslav-tmf committed:
Update Dockerfile to change working directory from /app to /src

Arsen-hrynevych commented on Dockerfile:
```diff
@@ -0,0 +1,15 @@
+FROM python:3.12-slim
+
+LABEL maintainer="vladislav.tmf@gmail.com"
+
+ENV PYTHONUNBUFFERED=1
+
+WORKDIR /src
+
+COPY requirements.txt requirements.txt
+
+RUN pip install --no-cache-dir -r requirements.txt
+
+COPY . .
+
+CMD ["python", "app/main.py"]
```
```
# syntax=docker/dockerfile:1

# Comments are provided throughout this file to help you get started.
# If you need more help, visit the Dockerfile reference guide at
# https://docs.docker.com/go/dockerfile-reference/

# Want to help us make this template better? Share your feedback here: https://forms.gle/ybq9Krt8jtBL3iCk7

ARG PYTHON_VERSION=3.12
FROM python:${PYTHON_VERSION}-slim as base

# Prevents Python from writing pyc files.
ENV PYTHONDONTWRITEBYTECODE=1

# Keeps Python from buffering stdout and stderr to avoid situations where
# the application crashes without emitting any logs due to buffering.
ENV PYTHONUNBUFFERED=1

WORKDIR /app

# Create a non-privileged user that the app will run under.
# See https://docs.docker.com/go/dockerfile-user-best-practices/
ARG UID=10001
RUN adduser \
    --disabled-password \
    --gecos "" \
    --home "/app" \
    --shell "/sbin/nologin" \
    --no-create-home \
    --uid "${UID}" \
    appuser

# Switch to the non-privileged user to run the application.
USER appuser

# Copy the source code into the container.
COPY ./app /app/

# Expose the port that the application listens on.
EXPOSE 8000

# Run the application.
CMD ["python", "app/main.py"]
```

Arsen-hrynevych commented on app/main.py:
```diff
@@ -1,6 +1,38 @@
+import os
+
+import requests
+
```
2 blank lines after imports

Arsen-hrynevych commented on app/main.py:
```diff
@@ -1,6 +1,38 @@
+
+import requests
+
+API_URL = "http://api.weatherapi.com/v1/current.json"
```
You need to generate API_KEY for using Weather API, but it shouldn't be hard-coded in main.py.

Arsen-hrynevych commented on app/main.py:
```diff
@@ -1,6 +1,38 @@
+import requests
+
+API_URL = "http://api.weatherapi.com/v1/current.json"
+CITY = "Paris"
```
it shouldn't be hard-coded in main.py.

Arsen-hrynevych commented on app/main.py:
```diff
@@ -1,6 +1,38 @@
+
 def get_weather() -> None:
-    # write your code here
-    pass
+    """
+    Fetches current weather data for Paris city and displays it in the console.
+    """
+    api_key = os.getenv("API_KEY")
+    if not api_key:
+        raise ValueError("API_KEY environment variable is not set")
+
+    params = {"key": api_key, "q": CITY}
+
+    response = requests.get(API_URL, params=params)
+    response.raise_for_status()
+
+    data = response.json()
+
+    location = data["location"]
+    country = location["country"]
+    current = data["current"]
+    local_time = location["localtime"]
+    current_temp = current["temp_c"]
+    current_conditions = current["condition"]["text"]
+
+    print(f"Performing request to Weather API for city {CITY}...")
+    print(
+        f"{CITY}/{country} {local_time} Weather: {current_temp} Celsius, "
+        f"{current_conditions}"
+    )
```
This doesn't follow the Single Responsibility Principle. You could split it into two functions and use a try-except block for each. Also, consider changing the naming of the response and other variables. Thanks!

Arsen-hrynevych reviewed (changes requested):
Please address the issues, and if you have any questions, write them next to the comments.

vladyslav-tmf committed:
Enhance weather application and improve Docker setup

- Updated .env.sample to include API_URL and CITY variables for flexible configuration.
- Refactored main.py to fetch weather data based on the CITY environment variable instead of a hardcoded value.
- Improved error handling in fetch_weather_data function.
- Modified Dockerfile to create a non-privileged user for running the application and changed the working directory to /app.

Arsen-hrynevych reviewed (approved):
Good
